src/Adapter/ContextStateManager.php

```diff
@@ -0,0 +1,192 @@
+<?php
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter;
+
+use Cart;
+use Context;
+use Country;
+use Currency;
+use Customer;
+use Language;
+
+/**
+ * Allows manipulating context state.
+ * This was adapted for specific broken legacy use cases when the previous state of context must be restored after some actions.
+ *
+ * e.g. order creation from back office.
+ *  Legacy requires Context properties (currency, country etc.) instead of using cart properties
+ *  so some context props must be changed for a while and then restored to previous state.
+ */
+final class ContextStateManager
+{
+    /**
+     * @var Context
+     */
+    private $context;
+
+    /**
+     * @var Cart|null
+     */
+    private $cart;
+
+    /**
+     * @var Country|null
+     */
+    private $country;
+
+    /**
+     * @var Currency|null
+     */
+    private $currency;
+
+    /**
+     * @var Language|null
+     */
+    private $language;
+
+    /**
+     * @var Customer|null
+     */
+    private $customer;
+
+    /**
+     * @var bool
+     */
+    private $cartStateChanged = false;
+
+    /**
+     * @var bool
+     */
+    private $countryStateChanged = false;
+
+    /**
+     * @var bool
+     */
+    private $currencyStateChanged = false;
+
+    /**
+     * @var bool
+     */
+    private $languageStateChanged = false;
+
+    /**
+     * @var bool
+     */
+    private $customerStateChanged = false;
+
+    /**
+     * @param Context $context
+     */
+    public function __construct(Context $context)
+    {
+        $this->context = $context;
+    }
+
+    /**
+     * Sets context cart and saves previous value
+     *
+     * @param Cart|null $cart
+     *
+     * @return $this
+     */
+    public function setCart(?Cart $cart): self
+    {
+        $this->cart = $this->context->cart;
```

Comment:
you should add a check:
```php
if (null === $this->savedCart) {
    $this->savedCart = $this->context->cart;
}
```
Because if the setter is called multiple times you will lose the reference on the initial context values

Reply:
good point. Just to check if cart was saved we should use `cartState` because NULL in context is also a valid value, so we cannot trust it.